conans/model/info.py

```diff
@@ -398,14 +401,36 @@ def default_std_matching(self):
         same as specifying None, packages are the same
         """
 
-        if self.full_settings.cppstd and \
-                self.full_settings.compiler and \
-                self.full_settings.compiler.version:
+        if self.full_settings.compiler and \
+            self.full_settings.compiler.version:
             default = cppstd_default(str(self.full_settings.compiler),
                                      str(self.full_settings.compiler.version))
-            if default == str(self.full_settings.cppstd):
+
+            if str(self.full_settings.cppstd) == default:
                 self.settings.cppstd = None
 
+            if str(self.full_settings.compiler.cppstd) == default:
+                def remove_cppstd(settings):
+                    try:
+                        settings.compiler.cppstd = None  # It's the default, assign None
+                    except AttributeError:
+                        # Your settings can be different at the moment of executing this function
+                        pass
+                    finally:
+                        return settings
+                self._std_matching = remove_cppstd
+
     def default_std_non_matching(self):
         if self.full_settings.cppstd:
             self.settings.cppstd = self.full_settings.cppstd
+        self._std_matching = None  # Do nothing
+
+    # Handle std_matching strategy for compiler.cppstd
+    _std_matching = None
```

Comment:
This variable here feels weird, why is it needed here and not at `__init__`?

Reply:
Maybe it reads easier if the var is always a function (empty lambda by default?) and it is called differently, something like: `self._adjust_settings_for_std`. You wouldn't need to write `if self._std_matching` just call always the `self._adjust_settings_for_std`.